keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,210 @@
+
+## Proposal
+
+The idea is to conduct the process of graduation to beta in small steps in the span of at least one Kubernetes release cycle. This will be done by creating one or more alpha versions of the config with the last alpha version being renamed to v1beta1 after the community is happy with it.
+Each of the sub-sections below can result in a separate alpha version release, although it will be better for users to have no more than a couple of alpha versions past v1alpha1.
+The proposed process is similar to the already successfully used one for kubeadm.
+
+### Re-encapsulate mode specific options
+
+The current state of the config has proven that:
+- Some options are deemed as mode specific, but are in fact shared between all modes.
+- Some options are placed directly into KubeProxyConfiguration, but are in fact mode specific ones.
+- There are options that are shared between some (but not all) modes. Specific features of the underlying implementation are common and this happens only within the boundaries of the platform (iptables and ipvs modes for example).
+- Although legacy Linux and Windows user mode proxies are separate code bases, they have a common set of options.
+
+With that in mind, the following measures are proposed:
+- Mode specific structs are consolidated to not use fields from other mode specific structs.
+- Introduce a single combined legacy user mode proxy struct for both Linux and Windows backends.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
+
+Platform specific fields need to be grouped together by platform name. This can be done by introducing structs to house them - one per platform (one for Linux, one for Windows, etc.). Then these structs will have an instance in the KubeProxyConfiguration.
+That can allow for users to specify settings for more than one platform per a YAML document, thus allowing for deduplication of platform independent shared fields. This can reduce the possibility of an error for users.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ... # modeA is platform independent
+platformX:
+  modeB: ...
+platformY:
+  modeC: ...
+```
+
+### Split instance specific and shared settings
+
+As a last step, instance specific and shared settings should be split into a couple of different YAML documents. This will allow for such settings to be specified in one or two YAML files, depending on user preferences.
+Preliminary reviews of the impacted settings show, that only settings, that are not platform or mode specific are instance specific. Therefore, the existing platform-mode specific structures are to be preserved in the shared settings.
+However, should the need arise, the platform-mode specific structures should be used in the instance specific structure as well.
+
+#### Example
+
+```yaml
+kind: KubeProxyInstanceConfiguration
```

Comment:
Again, this is pretty abstract - can we talk in concrete terms?  Which SPECIFIC fields do we see as instance?

@mtaufen at some point we argued about leaving instance-specific configs as flags, but I forget where we eventually settled?

Reply:
I have it updated with concrete fields (`HostnameOverride` and the `*BindAddress` fields). I also removed the mode specific stuff from there, as this is not applicable to the kube-proxy case, but was done more from an academic accuracy standpoint.
As far as the instance specific split is concerned, I am not fully sure about it. Certainly, we can enable instance specific (and probably shared too) fields to be overwritten from the command line. It's also certain, that some people may prefer the config file option for instance specifics. In that case a split is nice to have, as the shared config won't have to be patched (and therefore we can eliminate a possible source of errors).
However, this poses its own set of headaches - such as "What to do if we get a couple of InstanceConfigurations from different files?".

Reply:
@thockin as I remember it we were both ok with it as a stopgap but it's not really a long term solution if we want the interface to the component to be uniform (in terms of versioning and schema definition).

Reply:
> "What to do if we get a couple of InstanceConfigurations from different files?"

Is that likely to happen? Seems like something we should just reject.